### PR TITLE
Update the CLI upgrading topics

### PR DESCRIPTION
**Pull request by dshevtsov**
## Purpose of this pull request

This pull request adds an experimental scenario for CLI upgrade using the [plugin](https://github.com/magento/composer-root-update-plugin). This also splits the upgrading scenarios into separate topics.

## Affected DevDocs pages

- https://devdocs.magento.com/guides/v2.3/comp-mgr/cli/cli-upgrade.html

## Additional information

Internal ticket: MAGEDOC-3747

### CODE REVIEW

File: guides/v2.3/comp-mgr/cli/upgrade-with-plugin.md [hunk 1]
```diff
@@ -0,0 +1,195 @@
...
+2. Clear these directories:
+   - `var/cache`
+   - `var/page_cache`
+   - `generated/code`
+3. Check your storefront in your web browser again.
+
+<!-- Link definitions -->
+
+[file system ownership and permissions]: {{ page.baseurl }}/install-gde/prereq/file-system-perms.html
+[Prerequisites]: #prerequisites
```
Comment (jeff-matthews) [inline guides/v2.3/comp-mgr/cli/upgrade-with-plugin.md]: This anchor doesn't exist and is causing link failures in local previews.

File: guides/v2.3/comp-mgr/cli/upgrade-with-script.md [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+2. Clear these directories:
+   - `<magento_root>/var/cache`
+   - `<magento_root>/var/page_cache`
+   - `<magento_root>/generated/code`
+3. Check your storefront in your web browser again.
+
+<!-- Link definitions -->
+
+[file system ownership and permissions]: {{ page.baseurl }}/install-gde/prereq/file-system-perms.html
+[Prerequisites]: #prerequisites
```
Comment (jeff-matthews) [inline guides/v2.3/comp-mgr/cli/upgrade-with-script.md]: This anchor doesn't exist and is causing link failures in local previews.

File: _data/toc/software-update-guide.yml [hunk 3]
```diff
@@ -123,6 +123,10 @@ pages:
 
         - label: Command-line upgrade with sample data
           url: /comp-mgr/cli/cli-rc1-samp.html
+        - label: (Optional) Command-line upgrade using script
+          url: /comp-mgr/cli/upgrade-with-script.html
+        - label: (Experimental) Command-line upgrade using plugin
```
Comment (jeff-matthews) [inline _data/toc/software-update-guide.yml]: Should this be restricted to version 2.3?

File: guides/v2.3/comp-mgr/cli/cli-upgrade.md [hunk 4]
```diff
@@ -10,33 +10,25 @@ functional_areas:
...
-2. [Upgrade using the script] (semi-automated process)
-
-The upgrade scenario is the same for each of these options (both utilize Composer and a command line interface) from the system's point of view. However, if you upgrade using the script several of the steps are automated.
-
-{:.bs-callout .bs-callout-warning}
-Both scenarios require that you comply with the [Prerequisites].
-
-{:.bs-callout .bs-callout-info}
-If you cloned the Magento 2 GitHub repository, you **cannot** use this method to upgrade. Instead, see [Update the Magento application] for upgrade instructions.
+Do not use this method to upgrade, if you cloned the Magento 2 GitHub repository.
```
Comment (jeff-matthews) [inline guides/v2.3/comp-mgr/cli/cli-upgrade.md]:
```suggestion
Do not use this method to upgrade if you cloned the Magento 2 GitHub repository.
```

File: guides/v2.3/comp-mgr/cli/cli-upgrade.md [hunk 5]
```diff
@@ -10,33 +10,25 @@ functional_areas:
...
-* **Switch to maintenance mode**—To prevent access to your store while it's being upgraded, switch your store to maintenance mode:
+
+- **Complete the Update and upgrade checklist**—To avoid possible errors during installation or upgrading, complete the [Update and upgrade checklist].
+- **Set the `pub/` directory root**—If you set the Magento root directory to `pub/`, create another subdomain or docroot that uses the Magento installation directory as its root, and run the [System Upgrade utility] using that subdomain.
+- **Check PHP and environment settings**—Verify that your PHP and other environment settings are compatible with the [system requirements].
+- **Switch to maintenance mode**—To prevent access to your store while it's being upgraded, switch your store to maintenance mode:
 
     ```bash
-    php <magento_root>/bin/magento maintenance:enable
+    php bin/magento maintenance:enable
```
Comment (jeff-matthews) [inline guides/v2.3/comp-mgr/cli/cli-upgrade.md]:
```suggestion
    bin/magento maintenance:enable
```

File: guides/v2.3/comp-mgr/cli/cli-upgrade.md [hunk 6]
```diff
@@ -118,18 +116,19 @@ Open `composer.json` and edit the `"autoload": "psr-4"` section to include `"Zen
...
     },
-    ...
+    //...
 }
 ```
 
 ### Modify the Magento updater
 
-_Optional_—If the Magento updater is installed (it is located in `<Magento install dir>/update`) modify it:
+_Optional_—If the Magento updater is installed (it is located in `update/`) modify it:
```
Comment (jeff-matthews) [inline guides/v2.3/comp-mgr/cli/cli-upgrade.md]:
```suggestion
_Optional_—If the Magento updater is installed, modify it (it is located in the `update/` directory):
```

File: guides/v2.3/comp-mgr/cli/cli-upgrade.md [hunk 7]
```diff
@@ -118,18 +116,19 @@ Open `composer.json` and edit the `"autoload": "psr-4"` section to include `"Zen
...
 
-1. Backup `<Magento install dir>/update` directory.
-1. Create a Composer project.
+1. Backup `update/` directory.
```
Comment (jeff-matthews) [inline guides/v2.3/comp-mgr/cli/cli-upgrade.md]:
```suggestion
1. Backup the `update/` directory.
```

File: guides/v2.3/comp-mgr/cli/cli-upgrade.md [hunk 8]
```diff
@@ -142,21 +141,23 @@ _Optional_—If the Magento updater is installed (it is located in `<Magento ins
 
     {:.bs-callout .bs-callout-info}
       If you need to use a repository that contains non-public packages, such as internal sandboxes, change the URL in `--repository` accordingly.
-1. Remove the old `update` directory and move the `temp_dir/update` to the `<Magento install dir>/update` directory:
+3. Remove the old `update/` directory and move the `temp_dir/update/` to the `update/` directory:
```
Comment (jeff-matthews) [inline guides/v2.3/comp-mgr/cli/cli-upgrade.md]:
```suggestion
3. Remove the old `update/` directory and move `temp_dir/update/` to the `update/` directory:
```

File: guides/v2.3/comp-mgr/cli/cli-upgrade.md [hunk 9]
```diff
@@ -142,21 +141,23 @@ _Optional_—If the Magento updater is installed (it is located in `<Magento ins
...
    ```
+
    ```bash
    rm -rf temp_dir
    ```
 
 ### Update metadata
 
-Update the `"name"`, `"version"`, and `"description"` fields in the `<Magento install dir>/composer.json` file as desired.
+Update the `"name"`, `"version"`, and `"description"` fields in the `composer.json` file as desired.
```
Comment (jeff-matthews) [inline guides/v2.3/comp-mgr/cli/cli-upgrade.md]:
```suggestion
Update the `"name"`, `"version"`, and `"description"` fields in the `composer.json` file as needed.
```

File: guides/v2.3/comp-mgr/cli/cli-upgrade.md [hunk 10]
```diff
@@ -219,115 +222,38 @@ Open your storefront URL in a web browser to check whether the upgrade was succe
...
-   * `<magento_root>/var/page_cache`
-   * `<magento_root>/generated/code`
-1. Check your storefront in your web browser again.
-
-## Upgrade using the script {#upgrade-cli-script}
-
-Upgrading your Magento installation with our script, which makes the upgrade process semi-automated, is the preferred method—it's easy, quick, and efficient.
-
-Run this script after upgrading to PHP 7.1 or 7.2.
+2. Clear these directories:
```
Comment (jeff-matthews) [inline guides/v2.3/comp-mgr/cli/cli-upgrade.md]:
```suggestion
2. Clear the following directories:
```

File: guides/v2.3/comp-mgr/cli/cli-upgrade.md [hunk 11]
```diff
@@ -219,115 +222,38 @@ Open your storefront URL in a web browser to check whether the upgrade was succe
...
-
-Download, or copy, the script file from the `magento/magento2` repo, [`/dev/tools/UpgradeScripts/pre_composer_update_2.3.php`][script].
-
-### Show script options
+## Alternatives
 
-```bash
-php -f pre_composer_update_2.3.php -- --help
-```
+There are alternative ways that provide some automation to the upgrading process:
```
Comment (jeff-matthews) [inline guides/v2.3/comp-mgr/cli/cli-upgrade.md]:
```suggestion
There are alternatives methods that automate parts of the upgrade process:
```

File: guides/v2.3/comp-mgr/cli/cli-upgrade.md [hunk 12]
```diff
@@ -219,115 +222,38 @@ Open your storefront URL in a web browser to check whether the upgrade was succe
...
 
-### Run the script
+1. [Upgrade using the script][] (semi-automated process)
+   Upgrading using the script process is a bit easier and less intensive, if you have not made updates to values that the script affects.
```
Comment (jeff-matthews) [inline guides/v2.3/comp-mgr/cli/cli-upgrade.md]:
```suggestion
   Upgrading using the script process is a bit easier and less intensive if you have not made updates to values that the script affects.
```

File: guides/v2.3/comp-mgr/cli/cli-upgrade.md [hunk 13]
```diff
@@ -219,115 +222,38 @@ Open your storefront URL in a web browser to check whether the upgrade was succe
...
+   In case you previously made the updates, do not upgrade using the script, because this will override the affected values.
```
Comment (jeff-matthews) [inline guides/v2.3/comp-mgr/cli/cli-upgrade.md]:
```suggestion
   If you previously made updates, do not upgrade using the script. The script will override your updates.
```

File: guides/v2.3/comp-mgr/cli/cli-upgrade.md [hunk 14]
```diff
@@ -219,115 +222,38 @@ Open your storefront URL in a web browser to check whether the upgrade was succe
...
+2. EXPERIMENTAL: [Upgrade using the custom composer plugin][] (with a custom composer plugin)
```
Comment (jeff-matthews) [inline guides/v2.3/comp-mgr/cli/cli-upgrade.md]:
```suggestion
2. EXPERIMENTAL: [Upgrade using the custom Composer plugin][]
```

File: guides/v2.3/comp-mgr/cli/cli-upgrade.md [hunk 15]
```diff
@@ -219,115 +222,38 @@ Open your storefront URL in a web browser to check whether the upgrade was succe
...
+   We are developing a solution that would enhance the semi-automated process by a [custom composer plugin][] implementation.
```
Comment (jeff-matthews) [inline guides/v2.3/comp-mgr/cli/cli-upgrade.md]:
```suggestion
   We are developing a [custom Composer plugin][] that enhances the semi-automated upgrade process.
```

File: guides/v2.3/comp-mgr/cli/cli-upgrade.md [hunk 16]
```diff
@@ -219,115 +222,38 @@ Open your storefront URL in a web browser to check whether the upgrade was succe
...
 
-```bash
-php -f pre_composer_update_2.3.php -- --root='<path/to/magento/install/dir>' --repo=https://repo.magento.com/ <options>
-```
+The upgrading scenario is the same for each of these options (both utilize Composer and a command line interface) from the system point of view.
```
Comment (jeff-matthews) [inline guides/v2.3/comp-mgr/cli/cli-upgrade.md]:
```suggestion
The upgrading scenario is the same for each of these options. Both use Composer and a command line interface.
```

File: guides/v2.3/comp-mgr/cli/upgrade-with-plugin.md [hunk 17]
```diff
@@ -0,0 +1,195 @@
+---
+group: software-update-guide
+title: (Experimental) Command-line upgrade using the Magento composer root plugin
```
Comment (jeff-matthews) [inline guides/v2.3/comp-mgr/cli/upgrade-with-plugin.md]:
```suggestion
title: (Experimental) Upgrade using the Composer plugin
```

File: guides/v2.3/comp-mgr/cli/upgrade-with-plugin.md [hunk 18]
```diff
@@ -0,0 +1,195 @@
...
+<!-- Topic variables
+{% capture ce %}{{site.data.var.ce}}{% endcapture %}
+{% capture ee %}{{site.data.var.ee}}{% endcapture %}
+-->
+
+{: .bs-callout-warning}
+This upgrading scenario is still under development and is published here to make it available for the Community to test.
+Do not use it on production.
+
+The [`magento/composer-root-update-plugin`][custom composer plugin] Composer plugin resolves changes that need to be made to the root project `composer.json` file before updating to a new Magento product requirement.
```
Comment (jeff-matthews) [inline guides/v2.3/comp-mgr/cli/upgrade-with-plugin.md]: This link is not rendering properly in HTML.

File: guides/v2.3/comp-mgr/cli/upgrade-with-plugin.md [hunk 19]
```diff
@@ -0,0 +1,195 @@
...
+
+Basically, this section repeats the manual upgrade scenario with the only exclusion that you are guided to install the plugin to resolve the dependency conflicts instead of fixing them manually.
+
+## Ensure to comply with prerequisites
+
+Refer to [Prerequisites].
```
Comment (jeff-matthews) [inline guides/v2.3/comp-mgr/cli/upgrade-with-plugin.md]: Broken link

File: guides/v2.3/comp-mgr/cli/upgrade-with-plugin.md [hunk 20]
```diff
@@ -0,0 +1,195 @@
...
+
+{: .bs-callout-warning}
+This upgrading scenario is still under development and is published here to make it available for the Community to test.
+Do not use it on production.
+
+The [`magento/composer-root-update-plugin`][custom composer plugin] Composer plugin resolves changes that need to be made to the root project `composer.json` file before updating to a new Magento product requirement.
+
+Basically, this section repeats the manual upgrade scenario with the only exclusion that you are guided to install the plugin to resolve the dependency conflicts instead of fixing them manually.
+
+## Ensure to comply with prerequisites
```
Comment (jeff-matthews) [inline guides/v2.3/comp-mgr/cli/upgrade-with-plugin.md]:
```suggestion
## Before you begin
```

File: guides/v2.3/comp-mgr/cli/upgrade-with-plugin.md [hunk 21]
```diff
@@ -0,0 +1,195 @@
...
+   ```
+
+   ```bash
+   rm -rf temp_dir
+   ```
+
+## Update metadata
+
+{:.bs-callout .bs-callout-info}
+Updating the metadata in `composer.json` file is entirely superficial, not functional.
```
Comment (jeff-matthews) [inline guides/v2.3/comp-mgr/cli/upgrade-with-plugin.md]:
```suggestion
Updating the metadata in the `composer.json` file is entirely superficial, not functional.
```

Comment (jeff-matthews) [reply guides/v2.3/comp-mgr/cli/upgrade-with-plugin.md]: I'm not sure what this means precisely.

File: guides/v2.3/comp-mgr/cli/upgrade-with-plugin.md [hunk 22]
```diff
@@ -0,0 +1,195 @@
...
+
+Update the `"name"`, `"version"`, and `"description"` fields in the `composer.json` file as desired.
```
Comment (jeff-matthews) [inline guides/v2.3/comp-mgr/cli/upgrade-with-plugin.md]:
```suggestion
Update the `"name"`, `"version"`, and `"description"` fields in the `composer.json` file as needed.
```

File: guides/v2.3/comp-mgr/cli/upgrade-with-plugin.md [hunk 23]
```diff
@@ -0,0 +1,195 @@
...
+```bash
+rm -rf var/page_cache/*
+```
+
+```bash
+rm -rf generated/code/*
+```
+
+{:.bs-callout .bs-callout-info}
+If you use a cache storage other than the filesystem, such as Redis or Memcached, you must manually clear the cache there too.
```
Comment (jeff-matthews) [inline guides/v2.3/comp-mgr/cli/upgrade-with-plugin.md]:
```suggestion
If you use cache storage other than the filesystem, such as Redis or Memcached, you must manually clear the cache there too.
```

File: guides/v2.3/comp-mgr/cli/upgrade-with-script.md [hunk 24]
```diff
@@ -0,0 +1,108 @@
+---
+group: software-update-guide
+title: Command-line upgrade using the script
```
Comment (jeff-matthews) [inline guides/v2.3/comp-mgr/cli/upgrade-with-script.md]:
```suggestion
title: Upgrade using the custom script
```

File: guides/v2.3/comp-mgr/cli/upgrade-with-script.md [hunk 25]
```diff
@@ -0,0 +1,108 @@
...
+
+- Updates Magento with the 2.3 requirements.
+- Backs up the `composer.json` file.
+- Specifies the new version of the Magento metapackage.
+- Updates the `"require-dev"` section in the `composer.json` file.
+- Adds `"Zend\\Mvc\\Controller\\": "setup/src/Zend/Mvc/Controller/"` to the `"autoload":"psr-4"` section in the `composer.json` file.
+- Backs up and updates the `magento/updater`, if it has been installed.
+- Updates the `"name"`, `"version"`, and `"description"` fields in the `composer.json` file.
+
+## Ensure to comply with prerequisites
```
Comment (jeff-matthews) [inline guides/v2.3/comp-mgr/cli/upgrade-with-script.md]:
```suggestion
## Before you begin
```

File: guides/v2.3/comp-mgr/cli/upgrade-with-script.md [hunk 26]
```diff
@@ -0,0 +1,108 @@
...
+
+Refer to [Prerequisites].
```
Comment (jeff-matthews) [inline guides/v2.3/comp-mgr/cli/upgrade-with-script.md]: Broken link

File: guides/v2.3/comp-mgr/cli/upgrade-with-script.md [hunk 27]
```diff
@@ -0,0 +1,108 @@
...
+```bash
+rm -rf <Magento install dir>/var/page_cache/*
+```
+
+```bash
+rm -rf <Magento install dir>/generated/code/*
+```
+
+{:.bs-callout .bs-callout-info}
+If you use a cache storage other than the filesystem, such as Redis or Memcached, you must manually clear the cache there too.
```
Comment (jeff-matthews) [inline guides/v2.3/comp-mgr/cli/upgrade-with-script.md]:
```suggestion
If you use cache storage other than the filesystem, such as Redis or Memcached, you must manually clear the cache there too.
```

File: guides/v2.3/comp-mgr/cli/upgrade-with-plugin.md (hunk 1 above)
Comment (dshevtsov) [reply guides/v2.3/comp-mgr/cli/upgrade-with-plugin.md]: fixed

File: guides/v2.3/comp-mgr/cli/upgrade-with-script.md (hunk 2 above)
Comment (dshevtsov) [reply guides/v2.3/comp-mgr/cli/upgrade-with-script.md]: fixed

File: _data/toc/software-update-guide.yml (hunk 3 above)
Comment (dshevtsov) [reply _data/toc/software-update-guide.yml]: fixed

File: guides/v2.3/comp-mgr/cli/upgrade-with-script.md (hunk 26 above)
Comment (dshevtsov) [reply guides/v2.3/comp-mgr/cli/upgrade-with-script.md]: fixed

**Comment by dshevtsov** [comment]
@fascinosum please verify the values for the `composer require/remove --dev` values; they have changed in the releases since this doc was originally written for 2.3.0.

**Comment by jeff-matthews** [comment]
@dshevtsov, it doesn't look like you've accepted any of my editorial suggestions.

File: guides/v2.3/comp-mgr/cli/upgrade-with-plugin.md (hunk 18 above)
Comment (dshevtsov) [reply guides/v2.3/comp-mgr/cli/upgrade-with-plugin.md]: fixed

File: guides/v2.3/comp-mgr/cli/upgrade-with-plugin.md (hunk 19 above)
Comment (dshevtsov) [reply guides/v2.3/comp-mgr/cli/upgrade-with-plugin.md]: fixed

**Comment by jeff-matthews** [comment]
running tests

**Comment by jeff-matthews** [comment]
running tests

**Comment by jeff-matthews** [comment]
running tests

**Pull request merged**